File: MATTHEW/leetcode/AIcoding/5.Qwen-3/javascript_solutions/1074-number-of-submatrices-that-sum-to-target.js
function numSubmatrixSumTarget(target, matrix) {
    const rows = matrix.length;
    const cols = matrix[0].length;

    for (let r = 0; r < rows; r++) {
        for (let c = 1; c < cols; c++) {
            matrix[r][c] += matrix[r][c - 1];
        }
    }

    let count = 0;

    for (let c1 = 0; c1 < cols; c1++) {
        for (let c2 = c1; c2 < cols; c2++) {
            const map = new Map();
            map.set(0, 1);
            let sum = 0;

            for (let r = 0; r < rows; r++) {
                sum += matrix[r][c2] - (c1 > 0 ? matrix[r][c1 - 1] : 0);
                count += map.get(sum - target) || 0;
                map.set(sum, (map.get(sum) || 0) + 1);
            }
        }
    }

    return count;
}

File: leetcode/AIcoding/1.Gemini-2.5/go_solutions/1074-number-of-submatrices-that-sum-to-target.go
package main

func numSubmatrixSumTarget(matrix [][]int, target int) int {
	rows := len(matrix)
	cols := len(matrix[0])
	count := 0

	for c1 := 0; c1 < cols; c1++ {
		rowSums := make([]int, rows)
		for c2 := c1; c2 < cols; c2++ {
			for r := 0; r < rows; r++ {
				rowSums[r] += matrix[r][c2]
			}

			// Now rowSums is a 1D array, find subarrays that sum to target
			freq := make(map[int]int)
			freq[0] = 1
			currentSum := 0
			for _, val := range rowSums {
				currentSum += val
				if c, ok := freq[currentSum-target]; ok {
					count += c
				}
				freq[currentSum]++
			}
		}
	}

	return count
}